src/channels/spmc.go

```diff
@@ -0,0 +1,64 @@
+package channels
+
+import "sync"
+
+type SPMC[T any] struct {
+	input    <-chan T
+	output   []chan<- T
+	outputMx *sync.Mutex
+}
+
+func NewSPMC[T any](input <-chan T, blocking bool) *SPMC[T] {
+	channel := &SPMC[T]{
+		input:    input,
+		output:   make([]chan<- T, 0),
+		outputMx: &sync.Mutex{},
+	}
+
+	if blocking {
+		go channel.runBlocking()
+	} else {
+		go channel.run()
+	}
+
+	return channel
+}
+
+func (spmc *SPMC[T]) run() {
+	for {
+		payload := <-spmc.input
+
+		for _, channel := range spmc.output {
+			select {
+			case channel <- payload:
+			default:
+			}
+		}
+	}
+}
+
+func (spmc *SPMC[T]) runBlocking() {
+	for {
+		payload := <-spmc.input
+
+		for _, channel := range spmc.output {
+			channel <- payload
+		}
+	}
+}
+
+func (spmc *SPMC[T]) Add(output chan<- T) {
+	spmc.outputMx.Lock()
+	defer spmc.outputMx.Unlock()
+
+	spmc.output = append(spmc.output, output)
+}
+
+func (spmc *SPMC[T]) New(bufSize int) <-chan T {
```

Comment:
Why did you add the mutex? If its to avoid the concurrent iteration and modification on slice, shouldn't you lock the mutex on the iterations too?

Reply:
forgot to remove it